design/client-side-encryption_design.md

```diff
@@ -0,0 +1,554 @@
+# Proposal for client-side encryption
+
+## Abstract
+
+Currently only PersistentVolumes are stored in a Restic/Kopia repository, which means that only they are encrypted and the rest of the backup is not.
+It is desired that the cluster backup can be encrypted which overall will lead to better security.
+
+## Background
+
+As of today, Velero stores only PersistentVolume data in a Restic/Kopia repository.
+This repository is encrypted while the rest of the backup is not.
+Backups should be safe against illegal access and tampering.
+It is therefore desired to have the ability to store this data in an encrypted manner.
+
+## Goals
+
+- Encryption of the main cluster backup file with AES.
+- Basic configuration of the encryption, e.g. turning it on and off.
+- Ability to change encryption key for future backups
+
+## Non Goals
+
+- Other ciphers than AES (can be added later).
+- Encrypting anything other than the main backup file.
+- re-encrypt old backups if new encryption key was set
+
+## High-Level Design
+
+Encryption will happen during the backup process before the objects are persisted (e.g. written to an object store).
+Decryption will happen during a restore after the persisted objects have been read.
+The encryption key will be stored in a Kubernetes secret by the user.
```

Comment:
what are your thoughts on using cloud provider based secret stores for example KeyVault in Azure / AWS Secret manager for maintaing the secret.

Because we are all aware that kubernetes secrets are just plaintext and do no come with any security.

Additionally the lifecycle of the secret is tied to the lifecycle of the cluster.

Storing clusters in secret stores outside might be more secure and durable. Think cross cluster restore

Reply:
I understand that this might not be specific use case for you, but I want to bring this up in context of design extensibility. How can current design extend to support such scenarios.
or if it is not planned to supported at all, please put it in non goals

Reply:
I designed this feature explicitly in such a way that fetching secrets from a KMS can be implemented as well. See the `KeyRetriever` interface. Although, maybe this might have to be extended by a plugin API for it to be truly swappable.